src/app/clusters/fan-control-server/fan-control-server.cpp

```diff
@@ -346,8 +346,9 @@ void MatterFanControlClusterServerAttributeChangedCallback(const app::ConcreteAt
             VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
                            ChipLogError(Zcl, "Failed to get SpeedSetting with error: 0x%02x", status));
 
-            float percent        = percentSetting.Value();
-            uint8_t speedSetting = static_cast<uint8_t>(ceil(speedMax * (percent * 0.01)));
+            float percent = percentSetting.Value();
+            // Minus insignificant number 0.00000001 before ceil() to avoid floating point precision error
+            uint8_t speedSetting = static_cast<uint8_t>(ceil(speedMax * (percent * 0.01) - 0.00000001));
```

Comment:
why not use "round" then? the 0.000...01 seems very arbitrary. Why that and not call 0.1% arbitrary (like 0.001) ?

From my tests with speedMax 10, it looks like just `static_casth<uint8_t>(round(speedMax * percent * 0.01)) is fine.

Also maybe this should be:


```cpp
float percent = 0.01f * percentSetting.Value();
uint8_t speedSetting = static_cast<uint8_t>(round(speedMax * percent));  
```

Reply:
We cannot use round because we need to use `ceil()` defined in spec: 
![image](https://github.com/project-chip/connectedhomeip/assets/7869096/74d26ea1-171c-4ce2-b64a-0f4eca7dd8a5)

For example, if speedMax is 4 (levels), and we set to 10%
```
uint8_t speedSetting = static_cast<uint8_t>(round(speedMax * percent))
                                      = round(4*0.1)
                                      = round(0.4)
                                      = 0
```
This is incorrect because according to spec it has to be 1 (first level) instead of 0 (off).

The insignificant value could also be `0.0001` or `0.000001` since the precision error happens in very low decimal places (e.g., in this case it happens at 15th decimal). I choose `0.00000001` because this would be more insignificant and be more safe so that it won't affect the calculation result.

On the other hand, the floating point precision is a little unpredictable. In my situation it happens when `10*0.7` but not happens when `10*0.8`. (Maybe this may depends on gcc, OS, CPU, ..)

Reply:
I would opt for a number of 0s that I could count at a quick glance. 0.001 or 0.00001 is ok, more digits is starting being odd.

Since our percent is actually in units, I think it is fair to call 0.001 (i.e. 10 times less than the small usable step) as the insignificant adjustment factor. At least that seems less environment specific.

Reply:
Realistically, given that percent is an integer to start with, we could also just do, as uint32 to avoid overflow:
```
(SpeedMax * percent + 99)/100
```
to compute ceil(SpeedMax * percent * 0.01).  That's integer division, no floating point ops, no rounding issues.

Reply:
@bzbarsky-apple Good solution! Fixed accordingly. Thanks!